Add `erc20-token-stream` permission

jeffsmale90 opened the pull request:
# Add ERC20 Token Stream Permission Support

## Overview
This PR introduces support for ERC20 token streaming permissions.

## Key Changes
- Added new `erc20-token-stream` permission type with all supported components.
- Decomposed permission specific forms out of demo dapp page into individual components
- Updated addingNewPermissions.md documentation to guide developers extending the supported permissions.

## Note
- This permission does not yet show the permission icon
- Additional work is required to ensure security around the specified token (token alerts / token address) - UX input will be required to get this right

jeffsmale90 committed:
Minor improvements in preparation for erc-20 permissions
- Handle token metadata decoding errors
- Pass decimals for accurate balance formatting
- Remove unused functions
- Include `decimals` in `accountDetails` of `NativeTokenPeriodicContext` and `NativeTokenStreamContext`
- Update types for `balance` (to `Hex`) and add `decimals` in permission context definitions

jeffsmale90 committed:
Add new erc20-token-stream permission type

jeffsmale90 committed:
Add tests for erc20TokenStream permission

jeffsmale90 committed:
Fix linting problems

jeffsmale90 committed:
Remove rulestoAdd from metadatas

jeffsmale90 committed:
Decompose specific permission form out of site/pages/index.tsx to allow rapid addition of new permissions

jeffsmale90 committed:
Update test constants to include new permission in expected registry

V00D00-child reviewed (commented):
Overall, this looks good! Just a few comments; nothing major. There is quite a bit of redundancy, a few TODOs, and some patterns that we should address to maintain high quality and keep the codebase healthy.

jeffsmale90 committed:
Add tests for erc20-token-stream, update existing token context to use decimals, and not just assume ether

jeffsmale90 committed:
Move token metadata into tokenMetadata key on context objects

jeffsmale90 committed:
centralize permission context validation logic

Extracted and reused shared validation functions across token stream and periodic permission contexts. Updated docs formatting and test assertions to reflect revised validation messages.

jeffsmale90 committed:
Fetch token metadata (#99)

* Add accountApiClient and corresponding tests to fetch account balance and metadata

* Use accountApiClient within AccountController to fetch token metadata and balance

- Introduce 'getTokenBalanceAndMetadata' to account controllers
- Add placeholder icons for missing network/token SVGs
- Replace hardcoded ETH token details in UI with dynamic values

* Add fallback to onchain metadata for any chain other than mainnet. Don't even try to return iconUrl, because we can't do anything with it anyways

* Update native-token-periodic permission

* Fix linting problems

* Minor tidy ups to metadata client implementation

- Introduce ACCOUNT_API_BASE_URL to snap config and enforce presence
- Rename tokenMetadataClient to blockchainMetadataClient
- Improve dynamic chain handling and call encoding in token metadata client
- Add tests for BlockchainTokenMetadataClient

* Remove errant console.logs from debugging

* Update build-lint-test action
- Add ACCOUNT_API_BASE_URL
- Move configuration out of individual steps to avoid duplication

* Ignore iconUrl received from AccountAPI

* Decompose token balance and metadata out of account controller into TokenMetadataService.

* Fix linting problems in tests

jeffsmale90 committed:
Decompose permission context into BaseTokenPermissionContext, add tests for decimals. Address todos

jeffsmale90 committed:
Use crypto.getRandomValues for salt. Use helper utils to validate bigint values.

jeffsmale90 committed:
Lint

jeffsmale90 committed:
Remove direct call to console.error, update add new permission docs, add tests for validateHexInteger

7zkm7b8gw9-web reviewed (commented):
Can we get the rest of the missing or wrongly deleted information
